README

                        \\   /
                         \\ / Version 11 Rel. 6.4
                         / \\
                        /   \\  Server for MiNT/GEM
                      ===============================
                               Sep 05, 2001


Contents
========

	1. Requirements
	
	2. Installation
	   2.1. Preparing Environment
	   2.2. Preparing Directories
	
	3. Fonts
	   3.1. Handling of GEM-Fonts
	   3.2. Font-Aliasing
	
	4. Using The X Server
	   4.1. Menu Bar Items
	   4.2. Keyboard Input
	   4.3. Mouse Actions


1. Requirements
===============

To get the server running with a useable speed, a machine with a 68030 and
10 MByte of Ram is needed at least.  Absolutely vital installed and working
system programs are: FreeMiNT >= 1.15.6, NVDI >= 4.1, N.AES or XaAES or
AES 4.1.

If you haven't MiNTnet installed you need a very small package (I call it
'tiny-sock') from X11.freemint.de installed instead.
    
Screen size (width*height) isn't restricted, but ST-High (640*400) is the
minimum to get it useable.  Nearly all color depths are supported, expect
of 2 planes (4 colors) and 15 bit PC format.

The latest binary version can be downloaded from the homepage at
   http://X11.freemint.de ,
there is also a source snapshot.  The very latest sources are accessible
from the FreeMiNT CVS Repository at
   http://wh58-508.st.uni-magdeburg.de/cgi-bin/cvsweb/xg/
Bug reports should be mailed to <AltF4@freemint.de>.


2. Installation
===============

Copy the server's files to some location on your filesystem; there is no
requirement of a fixed path for the server.  A usefull place might be
/usr/X11R6/bin/ but everything else works as well.

The next step is to make a directory /var/lib/Xapp/ for the server.  If the
server is startet by root at the first time, or the directory is writeable
by the server, it will do the following steps by itself.  Else do the
following (as root user):
	cd /var/lib ; mkdir Xapp ; chmod 777 Xapp

2.1. Preparing Environment
--------------------------

There are a few things necessary for X-clients to be able to work.

If a client starts, it first parses the environment variable DISPLAY to
find out at which IP number the server should be connected.  So it's
possible to start clients from connected machines, having their X-Windows
opend on your local machine.  The value of DISPLAY has to be of the form
"<server-IP>:0.0".  For clients running at a local machine, DISPLAY
contains simply the value "0:0.0", where the first "0" is simply a shortcut
for the local IP or even "localhost".  For clients started on a remote
machine connected to an Atari with e.g. the IP 123.234.34.56, DISPLAY at
the remote machine must be then "123.234.34.56:0.0".

Most comfortable for the user is, to have somewhere in mint.cnf the line
	setenv DISPLAY 0:0.0
so the setting it can't be forgotten.  Otherway it would be necessary to
set it by hand in a shell window and start every client from that shell.

2.2. Preparing Directories
--------------------------

The following isn't necessary for the server itself but should be done for
client programs to have a common directory hirachy.

First there should be a directory /etc/X11/app-defaults/ which will later
contain default configuration files for several client programs.  These
configuration files are such with the suffix ".ad" in packages and needs to
be renamed without the suffix while copying to the app-defaults e.g.
"Xcalc.ad" to only "Xcalc".  Also the X server will search for optional
configuration fileas at /etc/X11/.

Next, a directory /usr/X11R6/ should be created as being the "XTOPDIR" and
a link /usr/X11 pointing to this directory as well, for historical reasons:
   cd /usr ; ln -s X11R6 X11
Furthermore, some clients are looking for their configuration files not at
/etc/X11/app-defaults/, but at /usr/X11R6/lib/X11/app-defaults/ (or at
/usr/X11/lib/X11/app-defaults/, therefor the link) instead, so there should
be placed a link pointing to /etc/X11/app-defaults/:
   cd /usr/X11R6/lib/X11 ; ln -s /etc/X11/app-defaults

After carrying out all the previous steps, the system should have these
directories and links:
   /var/lib/Xapp/
   /etc/X11/app-defaults/
   /usr/X11R6/lib/X11/app-defaults -> /etc/X11/app-defaults/
   /usr/X11                        -> X11R6/


3. Fonts
========

In X, fonts are named following the XLFD ('X Logical Font Description').
These font names are in fact a set of descriptive fields with '-' as
delimiters.  The format is as following:

   -fndry-fmly-wght-slnt-sWd-adStl-pxSz-ptSz-resx-resy-spc-avgWd-rgstry-enc

with the field meanings:

fndry:     Foundry is the name of the organization that supplied or last
           modified the font, e.g. Adobe, Bitstream.
fmly:      FamilyName of the font, e.g. Courier, Gothic, Helvetica
wght:      WeightName, e.g. Bold, Medim, Black
slnt:      Slant, a code-string with tha encoding: "R"=Roman, "I"=Italic,
           "O"=Oblique, aso.
sWd:       SetWidth, e.g. Normal, Condensed, Narrow
adStl:     AddStyle, e.g. Sans, Serif,
pxSz:      PixelSize of the body of the font, similar to the character cell
           in VDI
ptSz:      PointSize in decipoint, that is 1/10th point
resx,resy:   horizontal/vertical resolution for which the font was designed
spc:       Spacing, code-string to indicate the escapement, that is "P" for
           proportional and "M" or "C" for monospaced
avgWd:     AverageWidth of all characters in 1/20th pixels
rgstry,enc:   CharsetRegistry/Encoding, e.g. ISO8859-1

All of these fields my contain wildcards (a '*').  For example, often
clients requests a font named '-*-*-*-*-*-*-*-120-*-*-*-*-ISO8859-1' which
means some font of 12 point height and that charset encoding is needed, no
matter of which family or style and so on.  The server will than take the
first font where both criteria matches.

3.1. Handling of GEM-Fonts
--------------------------

At startup, the server parses all fonts provided by the VDI and translates
the available information to XLFD strings.  These are stored in the hash
file /var/lib/Xapp/fonts.db to accelerate further startups.

3.2. Font-Aliasing
------------------

If there exists a file /etc/X11/fonts.alias it will be read at server
startup and later be used to substitute font names from client requests to
fonts the server provides.  Have a look at the server's font hash file at
/var/lib/Xapp/fonts.db to see which fonts are provided by your system.
There is also an example configuration 'fonts.alias.EXMPL' coming along
with this file.

The format of the fonts.alias file is very simple, lines starting with a
'!' are comment lines and will be ignored.  The relevant lines are formed
the way "foo: bar" and "foo bar" and means in both cases: "find 'foo' and
replace it with 'bar'", but in different ways.  The first form (with the
colon) is used to substitute a _part_ of the font name and the second form
(without colon) is used to substitute the _whole_ font name.

The "foo: bar" form is usefull to replace particular fields of the font
name, for example the lines
   -adobe-:   -*-
   -courier-: -Courier10Pitch-
would cause that a request for '-adobe-courier-*-*-*-*-*-*-*-*-*-*-*-*'
would be translated to '-*-Courier10Pitch-*-*-*-*-*-*-*-*-*-*-*-*' which
matches the speedo font from the NVDI package.  The field delimeters '-'
should always be used (even if not forced) to avoid wrong results.

The "foo: bar" form is to substitute the whole font name and will be done
_AFTER_ the part substitution.

There are two often requested pseudo-fonts which should always be aliased:

fixed: A monospaced font that fits best the system capabilities, e.g. the
       font used the AES, or some other monospaced font of your choice.

variable: Some proportional font which should be of the same height as the
          font used for fixed.

To use the fonts from the NVDI package, it could looks like

fixed   -Bitstream-Courier10Pitch-Medium-I-Normal-Serif-16-90-0-0-C-70-ISO8859-1
variable -Bitstream-Life-Medium-R-Normal-Serif-16-90-0-0-P-68-ISO8859-1

The font to be substituted can also be a pattern, like

-adobe-times-*-*-*-*-*-*-*-*-*-*-iso8859-1 \
                                     -*-times*-medium-*-*-*-*-*-*-*-*-*-*-*
-*-Helvetica-Medium-R-Normal--*-120-*-*-*-*-*-*	\
                                  -*-gothic*-medium-r-*-*-*-120-*-*-*-*-*-*

Please pay attention that the '\' isn't allowed in the font.alias file;
it's only used here for better readability!  Instead both parts have to be
in one line, with spaces or tabs between.

There are other oftern used font names of the form '9x15', '6x13' and so
on, which means monospaced fonts of <width> * <height> pixel size.  If an
request of such a font name appeares and the server can't find it in its
XLFD list nor in the alias list, it takes a scalable font which can scaled
best to that geometry and creates a generic XLFD entry.


4. Using The X Server
=====================

After been started the X server checks if  /dev/xconout2  isn't locked by
some other program and if  /usr/X11/bin/xconsole  exists.  In this case the
latter on is started acts as a replacement for TosWin2's console window.
TosWin2 mustn't be terminated for having xconsole started, only closing its
console window is enough.  Note that also for xconsole the xconout2.xdd
must be installed.

The next step after this is checking if either  /etc/X11/Xmodmap  and 
/usr/X11/bin/xmodmap are existing.  In this case it will be started to
apply user defined keyboard and/or mouse button mappings.  For more
information see the example file 'Xmodmap.EXMPL' and the xmodmap manpage.

4.1. Menu Bar Items
-------------------

X11-Server:
	About:   Views version information.

File:
	gwm:   Enables/disables the built-in window manager.  In fact, this
		option controls if X windows have GEM windgets and decor, or not.
	Quit:   Quits the X server.  If some clients are connected, a
		confirmation dialog will apear first.

Clients:   Views a list of all client connections.  Selecting one of the
	entries while the Control key is held down, closes down immediately the
	corresponding connection.

4.2. Keyboard Input
-------------------

In opposite to GEM apllications, every X window gets keyboard events always
and only if it contains the mouse pointer, independent of the fact which is
the top window.  This means that even if some GEM application has some of
its wimdows the top window but the mouse pointer is over an X window, all
keyboard inout will be sent to the latter.  On the other hand, if the
X server has one of its windows the top window but the mouse pointer is
somewhere else than over an X window, all keyboard inputs will be
discarded.

Normaly the X server does no interpretition of any kyeboard input.  There
is one exception:  The combination Control-Alt-Backspace causes an
immediate server shutdown and is meant as a "Panic Key".

4.3. Mouse Actions
------------------

X windows can be resized by clicking on the frame around and dragging to
the new size.  Some windows may be set to be not resizeable, this can be
overridden by holding down the Control key.  A short mouseclick onto the
frame tops/bottoms the window.

Holding down the Alt key while clicking somewhere inside a window allows to
move it around.


ChangeLog
2001-09-21: 0.14
   - if the server can't load its RSC file the normal way, it also
     searches in /usr/X11/bin/ now.
   - added output buffer boundary check to event generating functions
     too, if lots of events happend the server was still crashing.
   - corrected calculation of output stream buffer, was taken too big.
   - some internal code tidy ups.
2001-09-18: 0.13.3
   - heaviely reworked handling of output stream buffer to avoid
     crashes due to buffer overflows, especially in GetImage and
     ListFonts request.
   - speeded up Get/PutImage request for color depth >= 16bit.
   - corrected drawing of window decor (border).
2001-09-13: 0.13.2
   - closed a memory leakage in GetImage request.
   - added a workaround for xoring monochrome bitmaps to a bitmap of
     >= 16bit color depth, to go round a bug in NVDI.
   - corrected return pixel values of 8bit packed pixel bitmaps.
2001-09-08: 0.13.1
   - implemented GetImage request.  now xmag (screen magnifier) and
     the 'AutoGrab Window' function of xv are working.
   - corrected RGB return values of QueryColors request, only the
     first value was set before.
   - changed colormap return value of GetWindowAttributes request to
     the default colormap, workaround for xv's 'AutoGrab Window'
     function.

2001-09-05: 0.13
   - implemented support for 16/24/32 bits color depth graphic modes.
   - fixed a tricky bug which was responsible for white drawings over
     white background on graphic depths >= 16 bits.
   - fixed a bug which crashes the AES if a client tryes to change the
     root window (desktop) background.
2001-09-02: 0.12.4
   - implemented support for Falcon HighColor graphic mode.
   - uses now a static VDI color map to speed up some color related
     requests and colormap changing.
   - always unload fonts of closed offscreen bitmaps if necessary, was
     sometimes forgotten before.
2001-08-27: 0.12.3
   - corrected handling of ServerUngrab request.
   - fixed a bug which caused the server hanging in an infinite loop
     while working off a WM_ONTOP event.
   - reset mouse pointer to normal GEM form if last X window is
     closed.
   - corrected handling of SIGCHLD in connection with xconsole.
2001-08-24: 0.12.2
   - fixed a bug which caused crashes after activating the built-in
     window manager.
   - avoid processes started by the server from becoming zombies after
     termination.
   - some smaller speed improvements.
2001-08-22: 0.12.1
   - fixed a bug in the calculation of clipping rectangels (clip-by-
     children mode) which was responsible for occasional undrawn
     window parts.
   - moved some spezialized AES/VDI call functions to the gemlib to
     use the optimized trap interfaces.  code adapted to the new
     gemlib version 0.42.

2001-07-31: 0.12
   - updated and expanded README and Xg.hyp files.  The Xg.hyp now
     contains also the manpages for xconsole and xmodmap.
   - if the built-in window manager get activated manually after a
     server reset or automatically at startup, xmodmap will be tried
     to be started to apply user defined keyboard and/or mouse button
     mappings (see Xmodmap.EXMPL file).
   - improved performing of server reset with respect to the built-in
     window manager (pseudo client).
   - implemented ChangeKeyboardMapping request and MappingNotify
     event.  this makes it possible to change the keyboard mapping
     for user's purpose, normally done with the 'xmodmap' tool.
   - implemented QueryKeymap request.
   - implemented UnmapSubwindows request, was simply forgotten in
     prior releases.
2001-07-27: 0.11.2
   - implemented GetKeyboardControl request for clients that want to
     know which keys has auto repeat enabled, like xkeycaps does.
   - improved handling of ALT+numblock inputs, nearly all characters
     defined in the Atari font can be typed in this way now.
   - completely reworked keyboard mapping (again), should be working
     now correctly with milan keyboards also, finally.
2001-07-16: 0.11.1
   - implemented GrabServer/UngrabServer requests.
   - reworked socket stream handling.  using an unified call interface
     for both accept and normal io events reduces the cpu usage of the
     server by aprox. 10% in standby.
   - changed gemlib trap interface to asm inline (not completed yet).
     this improves very little the binary size and gem calling speed
     but wasn't really worth the effort.

2001-07-13: 0.11
   - if the server starts up and founds that there is no console installed (or
     opened) it tryes to start xconsole instead.  this works only in case that
     the xconout2.xdd is installed.
   - created another latin 1 gdos-font '6x10' (in gdos-X-fonts-03.tgz), as
     beloved by xconsole and xterm.
   - implemented AV_START message handling to allow client programms to be
     started with corrected command lines.
   - implemented command line conversion from tos to unix path convention for
     client execution.  this makes it possible to start client programs simply
     from the desktop, where file parameters for the client are always given
     in dos convention.
2001-07-08: 0.10.3
   - fixed a stupid bug from last release which stopped window resizing due to
     not taking over the mouse control.
   - improved arguments for executing of client processes.  therefor windows are
     named correctly now, even if clients are started from desktop.
2001-07-07: 0.10.2
   - completely reworked keyboard mapping, should now work correctly with
     different keyboard layouts, also for french and swedish.  Additional,
     mostly can be made use of inputs done by ALT+numblock.
   - improved KeyRelease events timing, will simply be generated after the
     keyboard's initial repeat time expired.
   - re-activated Ctrl+Alt+Backspace for imediate server shutdown, was removed
     by accident.
   - fix a bug which caused the AES to freeze if one mouse button was pressed
     while another was still down.

src/Xapp.h
/*
 * resource set indices for xapp
 *
 * created by ORCS 2.18
 */

/*
 * Number of Strings:        133
 * Number of Bitblks:        1
 * Number of Iconblks:       77
 * Number of Color Iconblks: 3
 * Number of Color Icons:    5
 * Number of Tedinfos:       2
 * Number of Free Strings:   0
 * Number of Free Images:    0
 * Number of Objects:        141
 * Number of Trees:          4
 * Number of Userblks:       0
 * Number of Images:         155
 * Total file size:          14986
 */

#ifdef RSC_NAME
#undef RSC_NAME
#endif
#ifndef __ALCYON__
#define RSC_NAME "xapp"
#endif
#ifdef RSC_ID
#undef RSC_ID
#endif
#ifdef xapp
#define RSC_ID xapp
#else
#define RSC_ID 0
#endif

#ifndef RSC_STATIC_FILE
# define RSC_STATIC_FILE 0
#endif
#if !RSC_STATIC_FILE
#define NUM_STRINGS 133
#define NUM_FRSTR 0
#define NUM_UD 0
#define NUM_IMAGES 155
#define NUM_BB 1
#define NUM_FRIMG 0
#define NUM_IB 77
#define NUM_CIB 3
#define NUM_TI 2
#define NUM_OBS 141
#define NUM_TREE 4
#endif



#define MENU                               0 /* menu */
#define MENU_CLNT                          5 /* TITLE in tree MENU */
#define MENU_ABOUT                         8 /* STRING in tree MENU */
#define MENU_GWM                          17 /* STRING in tree MENU */
#define MENU_QUIT                         19 /* STRING in tree MENU */
#define MENU_CLNT_FACE                    20 /* BOX in tree MENU */
#define MENU_CLNT_FRST                    21 /* STRING in tree MENU */
#define MENU_CLNT_LAST                    50 /* STRING in tree MENU */

#define ABOUT                              1 /* form/dialog */
#define ABOUT_LOGO                         1 /* IMAGE in tree ABOUT */
#define ABOUT_OK                           3 /* BUTTON in tree ABOUT */
#define ABOUT_VERSN                        6 /* TEXT in tree ABOUT */
#define ABOUT_BUILD                        7 /* TEXT in tree ABOUT */

#define GLYPH                              2 /* form/dialog */
#define GLYPH_X_cursor                     1 /* ICON in tree GLYPH */
#define GLYPH_arrow                        2 /* ICON in tree GLYPH */
#define GLYPH_based_arrow_down             3 /* ICON in tree GLYPH */
#define GLYPH_based_arrow_up               4 /* ICON in tree GLYPH */
#define GLYPH_boat                         5 /* ICON in tree GLYPH */
#define GLYPH_bogosity                     6 /* ICON in tree GLYPH */
#define GLYPH_bottom_left_corner           7 /* ICON in tree GLYPH */
#define GLYPH_bottom_right_corner          8 /* ICON in tree GLYPH */
#define GLYPH_bottom_side                  9 /* ICON in tree GLYPH */
#define GLYPH_bottom_tee                  10 /* ICON in tree GLYPH */
#define GLYPH_box_spiral                  11 /* ICON in tree GLYPH */
#define GLYPH_center_ptr                  12 /* ICON in tree GLYPH */
#define GLYPH_circle                      13 /* ICON in tree GLYPH */
#define GLYPH_clock                       14 /* ICON in tree GLYPH */
#define GLYPH_coffee_mug                  15 /* ICON in tree GLYPH */
#define GLYPH_cross                       16 /* ICON in tree GLYPH */
#define GLYPH_cross_reverse               17 /* ICON in tree GLYPH */
#define GLYPH_crosshair                   18 /* ICON in tree GLYPH */
#define GLYPH_diamond_cross               19 /* ICON in tree GLYPH */
#define GLYPH_dot                         20 /* ICON in tree GLYPH */
#define GLYPH_dotbox                      21 /* ICON in tree GLYPH */
#define GLYPH_double_arrow                22 /* ICON in tree GLYPH */
#define GLYPH_draft_large                 23 /* ICON in tree GLYPH */
#define GLYPH_draft_small                 24 /* ICON in tree GLYPH */
#define GLYPH_draped_box                  25 /* ICON in tree GLYPH */
#define GLYPH_exchange                    26 /* ICON in tree GLYPH */
#define GLYPH_fleur                       27 /* ICON in tree GLYPH */
#define GLYPH_gobbler                     28 /* ICON in tree GLYPH */
#define GLYPH_gumby                       29 /* ICON in tree GLYPH */
#define GLYPH_hand1                       30 /* ICON in tree GLYPH */
#define GLYPH_hand2                       31 /* ICON in tree GLYPH */
#define GLYPH_heart                       32 /* ICON in tree GLYPH */
#define GLYPH_icon                        33 /* ICON in tree GLYPH */
#define GLYPH_iron_cross                  34 /* ICON in tree GLYPH */
#define GLYPH_left_ptr                    35 /* ICON in tree GLYPH */
#define GLYPH_left_side                   36 /* ICON in tree GLYPH */
#define GLYPH_left_tee                    37 /* ICON in tree GLYPH */
#define GLYPH_leftbutton                  38 /* ICON in tree GLYPH */
#define GLYPH_ll_angle                    39 /* ICON in tree GLYPH */
#define GLYPH_lr_angle                    40 /* ICON in tree GLYPH */
#define GLYPH_man                         41 /* ICON in tree GLYPH */
#define GLYPH_middlebutton                42 /* ICON in tree GLYPH */
#define GLYPH_mouse                       43 /* ICON in tree GLYPH */
#define GLYPH_pencil                      44 /* ICON in tree GLYPH */
#define GLYPH_pirate                      45 /* ICON in tree GLYPH */
#define GLYPH_plus                        46 /* ICON in tree GLYPH */
#define GLYPH_question_arrow              47 /* ICON in tree GLYPH */
#define GLYPH_right_ptr                   48 /* ICON in tree GLYPH */
#define GLYPH_right_side                  49 /* ICON in tree GLYPH */
#define GLYPH_right_tee                   50 /* ICON in tree GLYPH */
#define GLYPH_rightbutton                 51 /* ICON in tree GLYPH */
#define GLYPH_rtl_logo                    52 /* ICON in tree GLYPH */
#define GLYPH_sailboat                    53 /* ICON in tree GLYPH */
#define GLYPH_sb_down_arrow               54 /* ICON in tree GLYPH */
#define GLYPH_sb_h_double_arrow           55 /* ICON in tree GLYPH */
#define GLYPH_sb_left_arrow               56 /* ICON in tree GLYPH */
#define GLYPH_sb_right_arrow              57 /* ICON in tree GLYPH */
#define GLYPH_sb_up_arrow                 58 /* ICON in tree GLYPH */
#define GLYPH_sb_v_double_arrow           59 /* ICON in tree GLYPH */
#define GLYPH_shuttle                     60 /* ICON in tree GLYPH */
#define GLYPH_sizing                      61 /* ICON in tree GLYPH */
#define GLYPH_spider                      62 /* ICON in tree GLYPH */
#define GLYPH_spraycan                    63 /* ICON in tree GLYPH */
#define GLYPH_star                        64 /* ICON in tree GLYPH */
#define GLYPH_target                      65 /* ICON in tree GLYPH */
#define GLYPH_tcross                      66 /* ICON in tree GLYPH */
#define GLYPH_top_left_arrow              67 /* ICON in tree GLYPH */
#define GLYPH_top_left_corner             68 /* ICON in tree GLYPH */
#define GLYPH_top_right_corner            69 /* ICON in tree GLYPH */
#define GLYPH_top_side                    70 /* ICON in tree GLYPH */
#define GLYPH_top_tee                     71 /* ICON in tree GLYPH */
#define GLYPH_trek                        72 /* ICON in tree GLYPH */
#define GLYPH_ul_angle                    73 /* ICON in tree GLYPH */
#define GLYPH_umbrella                    74 /* ICON in tree GLYPH */
#define GLYPH_ur_angle                    75 /* ICON in tree GLYPH */
#define GLYPH_watch                       76 /* ICON in tree GLYPH */
#define GLYPH_xterm                       77 /* ICON in tree GLYPH */

#define ICONS                              3 /* form/dialog */
#define ICONS_BIG_X                        1 /* CICON in tree ICONS */
#define ICONS_SMALL_X                      2 /* CICON in tree ICONS */
#define ICONS_NAME_X                       3 /* CICON in tree ICONS */




#ifdef __STDC__
#ifndef _WORD
#  ifdef WORD
#    define _WORD WORD
#  else
#    define _WORD short
#  endif
#endif
extern _WORD xapp_rsc_load(_WORD wchar, _WORD hchar);
extern _WORD xapp_rsc_gaddr(_WORD type, _WORD idx, void *gaddr);
extern _WORD xapp_rsc_free(void);
#endif
